File: software-case-platform/mobile/lib/services/auth_service.dart
import '../core/api/api_client.dart';
import '../core/constants/api_constants.dart';
import '../models/auth_user.dart';
import 'storage/storage_service.dart';

class AuthService {
  final ApiClient apiClient;
  final StorageService storageService;

  AuthService({
    required this.apiClient,
    required this.storageService,
  });

  Future<AuthUser> login(String email, String password) async {
    // Si se usa modo demo o credenciales demo, permitir acceso rápido
    if (email == 'demo@barberia.com' && password == 'demo123') {
      final demoUser = AuthUser(
        id: 1,
        email: email,
        nombreCompleto: 'Administrador Barbershop Demo',
        rol: 'ADMIN',
        token: 'demo-mock-jwt-token',
      );
      apiClient.token = demoUser.token;
      await storageService.saveToken(demoUser.token);
      await storageService.saveUser(demoUser);
      return demoUser;
    }

    final response = await apiClient.post(ApiConstants.authLogin, {
      'email': email,
      'password': password,
    });

    String token = '';
    Map<String, dynamic> userMap = {};

    if (response is Map<String, dynamic>) {
      token = response['token'] ?? response['accessToken'] ?? '';
      userMap = response['usuario'] ?? response['user'] ?? response;
    }

    final authUser = AuthUser.fromJson(userMap, token);
    apiClient.token = token;
    await storageService.saveToken(token);
    await storageService.saveUser(authUser);
    return authUser;
  }

  Future<AuthUser> register(String email, String password, String nombreCompleto) async {
    final response = await apiClient.post(ApiConstants.authRegister, {
      'email': email,
      'password': password,
      'nombreCompleto': nombreCompleto,
    });

    String token = '';
    Map<String, dynamic> userMap = {};

    if (response is Map<String, dynamic>) {
      token = response['token'] ?? response['accessToken'] ?? '';
      userMap = response['usuario'] ?? response['user'] ?? response;
    }

    final authUser = AuthUser.fromJson(userMap, token);
    apiClient.token = token;
    await storageService.saveToken(token);
    await storageService.saveUser(authUser);
    return authUser;
  }

  Future<void> logout() async {
    apiClient.token = null;
    await storageService.clearAuth();
  }
}

File: software-case-platform/frontend/src/services/authService.ts
import { apiClient } from './api';

export interface UserProfile {
  id: number;
  nombreCompleto: string;
  email: string;
  rol: 'ADMIN' | 'INGENIERO' | 'ARQUITECTO';
  estado: string;
}

export interface AuthResponse {
  token: string;
  usuario: UserProfile;
  rol: string;
  tipoToken: string;
}

export const authService = {
  async login(email: string, password: string): Promise<AuthResponse> {
    const response = await apiClient.post<AuthResponse>('/api/auth/login', {
      email,
      password,
    });
    if (response.data.token) {
      localStorage.setItem('token', response.data.token);
      localStorage.setItem('user', JSON.stringify(response.data.usuario));
      if (response.data.usuario?.email) {
        localStorage.setItem('userEmail', response.data.usuario.email);
      }
      if (response.data.usuario?.nombreCompleto) {
        localStorage.setItem('userName', response.data.usuario.nombreCompleto);
      }
    }
    return response.data;
  },

  async register(
    nombreCompleto: string,
    email: string,
    password: string,
    rol: 'ADMIN' | 'INGENIERO' | 'ARQUITECTO' = 'ARQUITECTO'
  ): Promise<UserProfile> {
    const response = await apiClient.post<UserProfile>('/api/auth/register', {
      nombreCompleto,
      email,
      password,
      rol,
    });
    return response.data;
  },

  logout(): void {
    localStorage.removeItem('token');
    localStorage.removeItem('user');
    localStorage.removeItem('userEmail');
    localStorage.removeItem('userName');
  },

  getCurrentUser(): UserProfile | null {
    const userStr = localStorage.getItem('user');
    if (!userStr) return null;
    try {
      return JSON.parse(userStr);
    } catch {
      return null;
    }
  },

  getToken(): string | null {
    return localStorage.getItem('token');
  },

  isAuthenticated(): boolean {
    return !!localStorage.getItem('token');
  },
};
